frame/transaction-storage/src/lib.rs

```diff
@@ -0,0 +1,364 @@
+				),
+				Error::<T>::InvalidProof
+			);
+			Self::deposit_event(Event::ProofChecked);
+			Ok(().into())
+		}
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Stored data under specified index.
+		Stored(u32),
+		/// Renewed data under specified index.
+		Renewed(u32),
+		/// Storage proof was successfully checked.
+		ProofChecked,
+	}
+
+	/// Collection of extrinsic storage roots for the current block.
+	#[pallet::storage]
+	#[pallet::getter(fn transaction_roots)]
+	pub(super) type TransactionRoots<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::BlockNumber,
+		Blake2_128Concat,
+		u32,
+		TransactionInfo,
+		OptionQuery,
+	>;
+
+	/// Count indexed chunks for each block.
+	#[pallet::storage]
+	pub(super) type TransactionCount<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		T::BlockNumber,
+		u32,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn byte_fee)]
+	/// Storage fee per byte.
+	pub(super) type ByteFee<T: Config> = StorageValue<_, BalanceOf<T>>;
+
+	#[pallet::storage]
+	pub(super) type Counter<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	/// Storage period for data in blocks. Should match `sp_storage_proof::DEFAULT_STORAGE_PERIOD`
+	/// for block authoring.
+	#[pallet::storage]
+	pub(super) type StoragePeriod<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub byte_fee: BalanceOf<T>,
+		pub storage_period: T::BlockNumber,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self {
+				byte_fee: 10u32.into(),
+				storage_period: DEFAULT_STORAGE_PERIOD.into(),
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			<ByteFee<T>>::put(&self.byte_fee);
+			<StoragePeriod<T>>::put(&self.storage_period);
+		}
+	}
+
+	#[pallet::inherent]
+	impl<T: Config> ProvideInherent for Pallet<T> {
+		type Call = Call<T>;
+		type Error = InherentError;
+		const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
+
+		fn create_inherent(data: &InherentData) -> Option<Self::Call> {
+			let proof = data.get_data::<TransactionStorageProof>(&Self::INHERENT_IDENTIFIER).unwrap_or(None);
+			Some(Call::check_proof(proof))
+		}
+
+		fn check_inherent(_call: &Self::Call, _data: &InherentData) -> result::Result<(), Self::Error> {
+			Ok(())
+		}
+
+		fn is_inherent(call: &Self::Call) -> bool {
+			matches!(call, Call::check_proof(_))
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn apply_fee(origin: OriginFor<T>, size: u32) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let byte_fee = ByteFee::<T>::get().ok_or(Error::<T>::NotConfigured)?;
+			let fee = byte_fee.saturating_mul(size.into());
+			ensure!(T::Currency::can_slash(&sender, fee), Error::<T>::InsufficientFunds);
+			T::Currency::slash(&sender, fee);
```

Comment:
Should add a imbalance handler, which could allow you to send the fees to treasury, or burn them, or whatever other logic you can think of.

https://github.com/paritytech/substrate/blob/master/frame/treasury/src/lib.rs#L127